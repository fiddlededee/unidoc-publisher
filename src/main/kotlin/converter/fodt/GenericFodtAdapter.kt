package converter.fodt

import model.Node
import reader.HtmlNode
import reader.UnknownTagProcessing
import writer.OdtStyleList

interface GenericFodtAdapter {
    val unknownTagProcessingRule: HtmlNode.() -> UnknownTagProcessing
    fun basicStyle(): OdtStyleList
    fun Node.normalizeAll() {}
}