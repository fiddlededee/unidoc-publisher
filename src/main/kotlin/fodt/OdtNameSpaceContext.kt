package fodt

import javax.xml.namespace.NamespaceContext

class OdtNameSpaceContext : NamespaceContext {
    override fun getNamespaceURI(p0: String?): String? {
        val namespaceURI = when (p0) {
            "text" -> "urn:oasis:names:tc:opendocument:xmlns:text:1.0"
            "style" -> "urn:oasis:names:tc:opendocument:xmlns:style:1.0"
            "fo" -> "urn:oasis:names:tc:opendocument:xmlns:xsl-fo-compatible:1.0"
            "office" -> "urn:oasis:names:tc:opendocument:xmlns:office:1.0"
            "table" -> "urn:oasis:names:tc:opendocument:xmlns:table:1.0"
            "draw" -> "urn:oasis:names:tc:opendocument:xmlns:drawing:1.0"
            "svg" -> "http://www.w3.org/2000/svg"
            "loext" -> "urn:org:documentfoundation:names:experimental:office:xmlns:loext:1.0"
            else -> null
        }
        return namespaceURI
    }

    override fun getPrefix(p0: String?): String {
        throw UnsupportedOperationException()
    }

    override fun getPrefixes(p0: String?): MutableIterator<String> {
        throw UnsupportedOperationException()
    }
}

