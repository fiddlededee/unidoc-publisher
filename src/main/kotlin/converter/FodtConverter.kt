package converter

import common.GenericAdapter
import fodt.FodtGenerator
import model.Document
import model.Node
import model.Text
import org.redundent.kotlin.xml.PrintOptions
import reader.GenericHtmlReader
import reader.HtmlNode
import reader.UnknownTagProcessing
import writer.OdWriter
import writer.OdtStyleList

open class FodtConverter(init: FodtConverter.() -> Unit) {
    private var adapter: GenericAdapter? = null
    fun adaptWith(adapter: GenericAdapter) {
        unknownTagProcessingRule = adapter.unknownTagProcessingRule
        odtStyleList = adapter.basicStyle()
        this.adapter = adapter
    }

    var html: String? = null
        set(value) {
            val fieldNotNullable = value ?: throw Exception("html value can't be null")
            field = fieldNotNullable; htmlNode = HtmlNode(fieldNotNullable)
        }

    var xpath: String = "/html/body"
    var unknownTagProcessingRule: HtmlNode.() -> UnknownTagProcessing = { UnknownTagProcessing.UNDEFINDED }
    var customNodeProcessingRule: Node.(htmlNode: HtmlNode) -> Unit = {}
    var odtStyleList = OdtStyleList()
    var template: String? = null
    var ast: Node? = null
    var preList: ArrayList<PreRegion> = arrayListOf()
    var fodtGenerator: FodtGenerator? = null
    var fodt: String? = null
    private var htmlNode: HtmlNode? = null

    init {
        this.apply(init)
    }

    fun ast(): Node {
        return ast ?: throw Exception("AST is empty")
    }

    fun template(): String {
        return template ?: throw Exception("Template is empty")
    }

    fun preList(): ArrayList<PreRegion> {
        if (preList.isEmpty()) throw Exception("The list of pre FODT elements is empty")
        return preList
    }

    fun html(): String {
        return html ?: throw Exception("HTML is emplty")
    }

    fun fodt(): String {
        return fodt ?: throw Exception("Fodt is empty")
    }

    fun parse() {
        val localHtmlNode = htmlNode?.selectAtXpath(xpath)
            ?: throw Exception("Please set html")
        ast = Document()
        val localAst = ast ?: throw Exception("Error: ast variable was mutated")
        newReaderInstance(
            localAst, localHtmlNode,
            unknownTagProcessingRule, customNodeProcessingRule
        )
            .apply { localAst.setBasics(localHtmlNode); iterateAll() }
        // Don't understand html rendering rules, hope I'm right about this
        localAst.descendant { it is Text && it.text.isBlank() }.forEach { blankTextNode ->
            if (!blankTextNode.hasPrevious() ||
                !blankTextNode.previous().isInline ||
                blankTextNode.previous().roles.contains("br") ||
                !blankTextNode.hasNext() ||
                !blankTextNode.next().isInline ||
                blankTextNode.next().roles.contains("br")
            )
                blankTextNode.remove()
        }
    }

    fun generatePre() {
        val localAst = ast ?: throw Exception("Ast is null")
        adapter?.apply { localAst.normalizeAll() }
        (localAst.descendant { it.includeTags.isNotEmpty() } +
                localAst
                ).forEach { nodeToProcess ->
                val includeTags = nodeToProcess.includeTags
                    .let { if (it.isNotEmpty()) it.toSet() else setOf("all") }
                val pre = newOdWriterInstance(odtStyleList = odtStyleList)
                    .apply { (nodeToProcess.write(this)) }
                    .preOdNode.toString(PrintOptions(pretty = false))
                preList.add(PreRegion(includeTags, pre, nodeToProcess.isInline))
            }

    }

    fun generateFodt() {
        val localTemplate = template ?: throw Exception("Please, set template variable")
        if (preList.isEmpty()) throw Exception("Please set pre variable")
        fodtGenerator = FodtGenerator(preList, localTemplate)
        serializeFodt()
    }

    fun convert() {
        parse(); generatePre(); generateFodt()
    }

    fun ast2fodt() {
        generatePre(); generateFodt()
    }

    fun serializeFodt() {
        fodt = fodtGenerator?.serialize()
    }

    open fun newReaderInstance(
        ast: Node, htmlNode: HtmlNode,
        unknownTagProcessingRule: HtmlNode.() -> UnknownTagProcessing,
        customNodeProcessing: Node.(htmlNode: HtmlNode) -> Unit
    ): GenericHtmlReader {
        return GenericHtmlReader(ast, htmlNode, unknownTagProcessingRule, customNodeProcessing)
    }

    open fun newOdWriterInstance(odtStyleList: OdtStyleList): OdWriter {
        return OdWriter(odtStyleList = odtStyleList)
    }
}