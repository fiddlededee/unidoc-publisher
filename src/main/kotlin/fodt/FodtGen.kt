package fodt

import converter.PreRegion
import org.w3c.dom.Document
import org.w3c.dom.Element
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import org.xml.sax.InputSource
import java.io.StringReader
import java.io.StringWriter
import java.util.NoSuchElementException
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.OutputKeys
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult
import javax.xml.xpath.XPath
import javax.xml.xpath.XPathConstants
import javax.xml.xpath.XPathFactory

val styleNS = "urn:oasis:names:tc:opendocument:xmlns:style:1.0"
val foNS = "urn:oasis:names:tc:opendocument:xmlns:xsl-fo-compatible:1.0"

val xpathInstance: XPath = XPathFactory.newInstance().newXPath().apply {
    namespaceContext = OdtNameSpaceContext()
}


fun Document.serialize(): String {
    val transformerFactory = TransformerFactory.newInstance()
    val trans = transformerFactory.newTransformer()
    trans.setOutputProperty(OutputKeys.METHOD, "xml")
    val sw = StringWriter()
    val result = StreamResult(sw)
    val source = DOMSource(this.documentElement)
    trans.transform(source, result)
    return sw.toString()
}


fun String.parseStringAsXML(): Document {
    return DocumentBuilderFactory.newInstance()
        .apply { isNamespaceAware = true }
        .newDocumentBuilder()
        .parse(InputSource(StringReader(this)))
}


fun Node.xpath(xpath: String): NodeList {
    return xpathInstance.compile(xpath)
        .evaluate(this, XPathConstants.NODESET) as NodeList
}


fun NodeList.iterable(): Iterable<Node> = object : Iterable<Node> {
    override fun iterator(): Iterator<Node> = object : Iterator<Node> {
        private var index = 0
        override fun hasNext(): Boolean = index < this@iterable.length
        override fun next(): Node {
            if (!hasNext()) throw NoSuchElementException()
            return this@iterable.item(index++)
        }
    }
}

class FodtGenerator(val preFodtList: ArrayList<PreRegion>, template: String) {
    val enrichedTemplate = template.parseStringAsXML()
    private val automaticStyles: Document = DocumentBuilderFactory.newInstance().apply { isNamespaceAware = true }
        .newDocumentBuilder().parse(InputSource(StringReader("<root/>")))


    fun serialize(): String {
        return enrichedTemplate.serialize()
    }

    enum class StyleFamily { PARAGRAPH, TEXT, TABLE, TABLECELL, TABLECOLUMN, TABLEROW, GRAPHIC }

    var styleNum = 0
    fun setStyles(xpathExpression: String, styleFamily: StyleFamily, preFodt: Document) {
        val nodesWithStyles = preFodt.xpath(xpathExpression)
        nodesWithStyles.iterable().forEach {
            val styleNamePrefix = when (styleFamily) {
                StyleFamily.PARAGRAPH -> "PAMC"
                StyleFamily.TEXT -> "TAMC"
                StyleFamily.TABLE -> "TABAMC"
                StyleFamily.TABLECELL -> "TABCELLAMC"
                StyleFamily.TABLECOLUMN -> "TABCOLAMC"
                StyleFamily.TABLEROW -> "TABROWAMC"
                StyleFamily.GRAPHIC -> "GAMC"
            }
            val styleFamilyName = when (styleFamily) {
                StyleFamily.PARAGRAPH -> "paragraph"
                StyleFamily.TEXT -> "text"
                StyleFamily.TABLE -> "table"
                StyleFamily.TABLECELL -> "table-cell"
                StyleFamily.TABLECOLUMN -> "table-column"
                StyleFamily.TABLEROW -> "table-row"
                StyleFamily.GRAPHIC -> "graphic"
            }
            styleNum += 1
            val generatedStyleName = "$styleNamePrefix$styleNum"
            val automaticStylesStyle =
                automaticStyles.createElementNS(styleNS, "style:style")
            automaticStylesStyle.apply {
                setAttributeNS(styleNS, "style:name", generatedStyleName)
                setAttributeNS(styleNS, "style:family", styleFamilyName)
            }
            automaticStyles.documentElement.appendChild(automaticStylesStyle)
            val textPropertiesNodes = it.childNodes.iterable().filter { node ->
                node is Element && node.nodeName == "style:text-properties"
            }
            if (textPropertiesNodes.isNotEmpty()) {
                val textPropertiesNode = textPropertiesNodes[0]
                automaticStyles.adoptNode(textPropertiesNode)
                automaticStylesStyle.appendChild(textPropertiesNode)
            }
            val paragraphPropertiesNodes = it.childNodes.iterable().filter { node ->
                node is Element && node.nodeName == "style:paragraph-properties"
            }
            if (paragraphPropertiesNodes.isNotEmpty()) {
                val paragraphPropertiesNode = paragraphPropertiesNodes[0]
                automaticStyles.adoptNode(paragraphPropertiesNode)
                automaticStylesStyle.appendChild(paragraphPropertiesNode)
            }
            val graphicPropertiesNodes = it.childNodes.iterable().filter { node ->
                node is Element && node.nodeName == "style:graphic-properties"
            }
            if (graphicPropertiesNodes.isNotEmpty()) {
                val graphicPropertiesNode = graphicPropertiesNodes[0]
                automaticStyles.adoptNode(graphicPropertiesNode)
                automaticStylesStyle.appendChild(graphicPropertiesNode)
            }
            val tablePropertiesNodes = it.childNodes.iterable().filter { node ->
                node is Element && (node.nodeName == "style:table-properties"
                        || node.nodeName == "style:table-column-properties"
                        || node.nodeName == "style:table-cell-properties"
                        || node.nodeName == "style:table-row-properties")
            }
            if (tablePropertiesNodes.isNotEmpty()) {
                val tablePropertiesNode = tablePropertiesNodes[0]
                automaticStyles.adoptNode(tablePropertiesNode)
                automaticStylesStyle.appendChild(tablePropertiesNode)
            }
            val styleMasterPageName = (it as Element).getAttributeNode("style:master-page-name")
            if (styleMasterPageName != null) {
                automaticStylesStyle.apply {
                    setAttributeNS(
                        styleNS,
                        "style:master-page-name",
                        styleMasterPageName.nodeValue.replace(" ", "_20_")
                    )
                }
            }
            if (arrayOf(StyleFamily.TABLE, StyleFamily.TABLECELL, StyleFamily.TABLECOLUMN, StyleFamily.TABLEROW)
                    .contains(styleFamily)
            ) {
                it.setAttribute("table:style-name", generatedStyleName)
            } else {
                val styleNameAttribute =
                    if (styleFamily == StyleFamily.GRAPHIC)
                        it.getAttributeNode("draw:style-name") else
                        it.getAttributeNode("text:style-name")
                if (styleNameAttribute != null) {
                    automaticStylesStyle.apply {
                        setAttributeNS(
                            styleNS,
                            "style:parent-style-name",
                            styleNameAttribute.nodeValue.replace(" ", "_20_")
                        )
                    }
                    styleNameAttribute.nodeValue = generatedStyleName
                } else {
                    if (styleFamily == StyleFamily.GRAPHIC)
                        it.setAttribute("draw:style-name", generatedStyleName) else
                        it.setAttribute("text:style-name", generatedStyleName)
                }
            }
        }
    }

    fun processTaggedRegion(preRegion: PreRegion) {
        val preFodt = preRegion.pre.parseStringAsXML()
        setStyles("(//text:p|//text:h)", StyleFamily.PARAGRAPH, preFodt)
        setStyles("(//text:span)", StyleFamily.TEXT, preFodt)
        setStyles("(//table:table)", StyleFamily.TABLE, preFodt)
        setStyles("(//table:table-cell)", StyleFamily.TABLECELL, preFodt)
        setStyles("(//table:table-column)", StyleFamily.TABLECOLUMN, preFodt)
        setStyles("(//table:table-row)", StyleFamily.TABLEROW, preFodt)
        setStyles("(//draw:frame)", StyleFamily.GRAPHIC, preFodt)
        val officeText = enrichedTemplate.xpath("//office:body/office:text")
        if (officeText.length == 1) {
            val officeTextElement = officeText.item(0)
            officeTextElement.xpath("//text:variable-set[@text:name = 'include']")
                .iterable().map { it as Element }.forEach { markedElement ->
                    val tagName = markedElement.textContent
                    if (!preRegion.includeTags.contains(tagName)) return@forEach
                    val elementToReplace = if (preRegion.inLine) markedElement else
                        markedElement.parentNode
                    preFodt.xpath("/root/*").iterable().forEach {
                        enrichedTemplate.adoptNode(it)
                        elementToReplace.parentNode.insertBefore(it, elementToReplace)
                    }
                    elementToReplace.parentNode.removeChild(elementToReplace)
                }
            officeTextElement.xpath("//text:variable-set[@text:name = 'process']")
                .iterable().map { it as Element }.forEach {
                    if (it.textContent != "end") return@forEach
                    var deleteFromNode = it.parentNode
                    while (deleteFromNode.nextSibling != null) {
                        val nextSibling = deleteFromNode.nextSibling
                        deleteFromNode.parentNode.removeChild(deleteFromNode)
                        deleteFromNode = nextSibling
                    }
                    deleteFromNode.parentNode.removeChild(deleteFromNode)
                }
        } else {
            throw Exception("office:body/office:text not found")
        }

    }

    init {
        preFodtList.forEach { preRegion ->
            processTaggedRegion(preRegion)
        }

        val officeAutomaticStyles = enrichedTemplate.xpath("//office:automatic-styles")
        if (officeAutomaticStyles.length == 1) {
            val officeAutomaticStylesElement = officeAutomaticStyles.item(0)
            automaticStyles.xpath("/root/*").iterable().forEach {
                enrichedTemplate.adoptNode(it)
                officeAutomaticStylesElement.appendChild(it)
            }
        } else {
            throw Exception("office:automatic-styles not found")
        }
    }

}

