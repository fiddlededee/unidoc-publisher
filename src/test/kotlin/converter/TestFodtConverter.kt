package converter

import converter.fodt.FodtConverter
import org.intellij.lang.annotations.Language
import org.junit.jupiter.api.Test
import verify


class TestFodtConverter {
    @Language("XML")
    val testTemplate =
        """<office:document
                xmlns:office="urn:oasis:names:tc:opendocument:xmlns:office:1.0"
                xmlns:text="urn:oasis:names:tc:opendocument:xmlns:text:1.0"
                xmlns:style="urn:oasis:names:tc:opendocument:xmlns:style:1.0"
                xmlns:fo="urn:oasis:names:tc:opendocument:xmlns:xsl-fo-compatible:1.0">
                <office:automatic-styles>
                    <this-tag-should-be-left/>
                </office:automatic-styles>
                <office:body>
                    <office:text text:use-soft-page-breaks="true">
                        <this-tag-should-be-left/>
                            <text:p><text:variable-set text:name="include" office:value-type="string">all</text:variable-set></text:p>
                        <this-tag-should-be-left/>
                    </office:text>
                </office:body>
        </office:document>
            """



    @Test
    fun testFull() {
        FodtConverter {
            html = "<div><p>Some paragraph</p></div>"
            xpath = "/html/body/div"
            parse()
            generatePre()
            template = testTemplate
            generateFodt()
            fodt!!.verify()
        }
    }
}