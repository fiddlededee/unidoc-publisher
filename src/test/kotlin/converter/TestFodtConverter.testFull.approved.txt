<?xml version="1.0" encoding="UTF-8"?><office:document xmlns:fo="urn:oasis:names:tc:opendocument:xmlns:xsl-fo-compatible:1.0" xmlns:office="urn:oasis:names:tc:opendocument:xmlns:office:1.0" xmlns:style="urn:oasis:names:tc:opendocument:xmlns:style:1.0" xmlns:text="urn:oasis:names:tc:opendocument:xmlns:text:1.0">
                <office:automatic-styles>
                    <this-tag-should-be-left/>
                <style:style style:family="paragraph" style:name="PAMC1" style:parent-style-name="Text_20_body"/></office:automatic-styles>
                <office:body>
                    <office:text text:use-soft-page-breaks="true">
                        <this-tag-should-be-left/>
                            <text:p text:style-name="PAMC1">Some paragraph</text:p>
                        <this-tag-should-be-left/>
                    </office:text>
                </office:body>
        </office:document>
